classes/core/PKPRouter.inc.php

```diff
@@ -335,6 +336,38 @@ function url($request, $newContext = null, $handler = null, $op = null, $path =
 		assert(false);
 	}
 
+	/**
+	 * Build a URL to an endpoint in the API
+	 *
+	 * This method builds the correct URL depending on whether disable_path_info
+	 * and restful_urls are enabled in the config file.
+	 *
+	 * @param Request $request
+	 * @param string $contextPath
+	 * @param string $apiVersion
+	 * @param string $baseEndpoint Example: 'submissions'
+	 * @param string $endpointParams Example: '1', '1/galleys'
+	 * @return string
+	 */
+	public function getApiUrl($request, $contextPath, $apiVersion, $baseEndpoint = '', $endpointParams = '') {
```

Comment:
Is this a parallel implementation of the current API URL builder?

Reply:
(Asking because I think there are some escaping concerns missed here)

Reply:
What's the current API URL builder?

Reply:
Just a reminder here: is there another method for building an API URL that I should be using here?

Reply:
So my memory proves OK -- I actually hadn't dug into this aspect of the API yet!

Essentially, here's how it all hangs together:

We have 3 different "router" implementations:
- `lib/pkp/classes/core/APIRouter.inc.php`
- `classes/core/PageRouter.inc.php`
- `lib/pkp/classes/core/PKPComponentRouter.inc.php`

Calls to `Request::url` (and, by extension, Smarty `{url ...}` calls) are delegated to these via the `Router` class. Normally it's not necessary to specify in `Request::url` or `{url ...}` calls which router is intended -- in general page requests use the page router, and component requests use the component router -- but in some cases you'll see it specified in template files using the `router` attribute (see e.g. [`templates/common/header.tpl`](https://github.com/pkp/pkp-lib/blob/a3d4022f0654c3d531ba79e46bdf140551d1c07c/templates/common/header.tpl#L38)).

The main benefit of this approach is that *all* URL generation goes through one pipeline, so all the various "modes" of URL generation don't get scattered throughout the codebase. Those are
- `path_info_disabled` (`config.inc.php`)
- `restful_urls` (`config.inc.php`)
- `base_url[xyz]` (`config.inc.php`)
- Hook-based intervention (this hasn't been fleshed out or required yet)

In particular, the `path_info_disabled` setting has historically meant we've strenuously avoided manual concatenation of URLs in favour of `Request::url` / `{url ...}`.

Grepping around, the same approach is available for API URLs, but hasn't been implemented or used. This would essentially mean implementing a `url` function in `APIRouter`, and specifying `router=$smarty.const.ROUTE_API` in calls to `{url ...}`. When providing URLs to Javascript, we've used slugs, e.g. `GIVEN_NAME_PLACEHOLDER` in [`createReviewerForm.tpl`](https://github.com/pkp/pkp-lib/blob/a3d4022f0654c3d531ba79e46bdf140551d1c07c/templates/controllers/grid/users/reviewer/form/createReviewerForm.tpl#L17), which is then replaced in Javascript using `String.replace` (e.g. [`UserFormHandler.js`](https://github.com/pkp/pkp-lib/blob/a3d4022f0654c3d531ba79e46bdf140551d1c07c/js/controllers/form/UserFormHandler.js#L103)). This permits variable replacement in JS-land while still keeping all URL generation on the PHP/Smarty side, and should prevent either having to concatenate URLs throughout the codebase, or having to maintain a parallel implementation of the URL generation code in Javascript.

It looks like the current API URL generation is an outlier -- manual URL concatenation is used. I'm open to ideas and not sure what the implications are e.g. in `vue.js` but I think we should keep URL generation to PHP-land if possible.

Reply:
Ah ok! I avoided the `PKPRouter::url()` method because I thought it would be awkward to try to shoehorn Slim's URL structure into our own url generation (there's no `op`, `path` or `anchor` for API urls, just an endpoint). I've worked around this by converting `$handler` to `$endpoint`. Does this look alright?

https://github.com/NateWr/pkp-lib/commit/1da32440aaf1d3deb746d72a9e98ea26ee28c38e
https://github.com/NateWr/ojs/commit/b4a15da77afd4ede8091609a9d37ef50f802fe44

(While testing, it's become apparent that only `GET` requests work with path info disabled. I believe that's the case in `master` too, but I'm working on fixing that now.)

Reply:
Yes, that looks much better -- thanks, @NateWr! I had thought there were cases of suffixes getting glued onto partial URLs in the JS side that I don't see here in the PRs, but you know that code better than I do...